include/v4l2_camera/rate_bound_status.hpp

```diff
@@ -0,0 +1,281 @@
+  };
+
+  struct Ok : public StateBase
+  {
+    Ok()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::OK,
+                    "Rate is reasonable") {}
+  };
+
+  struct Warn : public StateBase
+  {
+    Warn()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN,
+                    "Rate is within warning range") {}
+  };
+
+  struct Error : public StateBase
+  {
+    Error()
+        : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR,
+                    "Rate is out of valid range") {}
+  };
+
+  using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+ public:
+  /**
+   * \brief Constructs RateBoundstatus, which inherits diagnostic_updater::DiagnosticTask.
+   *
+   * \param ok_params The pair of min/max frequency for the topic rate to be recognized as "OK".
+   * \param warn_params The pair of min/max frequency for the topic rate to be recognized as "WARN".
+   * These values should have a wider range than `ok_params`.
+   * \param num_frame_transition The number of the successive observations for the status
+   * transition. E.g., the status will not be changed from OK to WARN until successive
+   * `num_frame_transition` WARNs are observed.
+   * \param immediate_error_report If true (default), errors related to the rate bounds will be
+   * reported immediately once it is observed; otherwise, hysteresis manner using
+   * `num_frame_transition` will be adopted
+   * \param name The arbitral string to be assigned for this diagnostic task.
+   * This name will not be exposed in the actual published topics.
+   */
+  RateBoundStatus(const RateBoundStatusParam& ok_params,
+                  const RateBoundStatusParam& warn_params,
+                  const size_t num_frame_transition = 1,
+                  const bool immediate_error_report = true,
+                  std::string name = "rate bound check")
+      : DiagnosticTask(name), ok_params_(ok_params), warn_params_(warn_params),
+        num_frame_transition_(num_frame_transition),
+        immediate_error_report_(immediate_error_report), zero_seen_(false),
+        candidate_state_(Stale{}), current_state_(Stale{})
+  {
+    if (num_frame_transition < 1) {
+      num_frame_transition_ = 1;
+    }
+
+    // Confirm `warn_params` surely has wider range than `ok_params`
+    if (!(warn_params_.min_frequency < ok_params_.min_frequency &&
+          ok_params_.max_frequency< warn_params_.max_frequency)) {
+      throw std::runtime_error(
+          "Invald range parameters were detected. warn_params should specify a range "
+          "that includes a range of ok_params.");
+    }
+  }
+
+  /**
+   * \brief Signals an event.
+   *
+   * \param t The timestamp of the event that will be used in computing
+   * intervals.
+   */
+  void tick(double stamp)
+  {
+    std::unique_lock<std::mutex> lock(lock_);
+    if (!previous_frame_timestamp_) {
+      previous_frame_timestamp_ = rclcpp::Clock().now().seconds();
```

Comment:
Could maybe consider constructing the clock in the class constructor and sharing it - minor, but if we have many diagnostics all constructing a clock every frame, and some topics may be very high frequency.

Reply:
Due to the change of 5b9e000, now we are not able to have an instance of a clock because (AFAIK) `std::chrono::steady_clock` does not provide instantiation. Some investigation results say this steady_clock may cause relatively high overhead for calling, but I consider that this is inevitable.